File: src/caps.ts
/**
 * Named-capability versioning for client/server compatibility.
 *
 * The server advertises a `caps` object on `GET /api/server-info` mapping each
 * compatibility surface to an opaque string version. The client intersects
 * those values against `REQUIRED_CAPS` below; on any mismatch the client
 * hard-blocks connect/sync rather than proceeding and corrupting state.
 *
 * Cap values are bumped **only** on a wire-incompatible change to that surface
 * (removing/renaming a field, changing a type, changing semantics). Adding an
 * optional field does NOT bump. See `AGENTS.md` "Compatibility & versioning".
 *
 * Only the Obsidian plugin enforces caps. The SDK and CLI mirror the optional
 * `caps`/`requiredCaps` fields on `ServerInfoResponse` so consumers can
 * self-gate, but they do not block on mismatches themselves.
 */

import { SYNC_FORMAT } from "./pluginDb/types";

/** Names of every compatibility surface the plugin understands. */
export const CAP_NAMES = [
  "restApi",
  "oauth",
  "pluginDbSync",
  "attachmentShim",
  "documentEpoch",
  "documentInvalidation",
] as const;
export type CapName = (typeof CAP_NAMES)[number];

/**
 * Accepted cap values per surface. A server advertising a value not in this
 * list is treated as incompatible (direction cannot be inferred from opaque
 * strings, so the result is the neutral `"server-incompatible"`).
 *
 * To accept a range of server releases during a migration window, list multiple
 * values here (e.g. `restApi: ["1", "2"]`).
 */
export const REQUIRED_CAPS: Record<CapName, string[]> = {
  restApi: ["3"],
  oauth: ["1"],
  pluginDbSync: [SYNC_FORMAT],
  attachmentShim: ["https://realtime.md/attachment-shim/v1"],
  documentEpoch: ["1"],
  documentInvalidation: ["1"],
};

/**
 * Caps required for every connection. `documentInvalidation` is intentionally
 * optional: clients can still sync with an older server, but must retain every
 * mobile document in memory because that server cannot wake an evicted child.
 */
const MANDATORY_CAPS: readonly CapName[] = [
  "restApi",
  "oauth",
  "pluginDbSync",
  "attachmentShim",
  "documentEpoch",
];

/** Whether a server advertises an accepted version of a known capability. */
export function serverSupportsCapability(caps: unknown, name: CapName): boolean {
  if (caps === null || caps === undefined || typeof caps !== "object" || Array.isArray(caps)) {
    return false;
  }
  const advertised = (caps as Record<string, unknown>)[name];
  return typeof advertised === "string" && REQUIRED_CAPS[name].includes(advertised);
}

export type CompatibilityResult =
  | { ok: true }
  | { ok: false; reason: "server-incompatible" | "client-too-old"; detail: string };

/**
 * Check a server's advertised `caps`/`requiredCaps` against the client's
 * accepted values. Pure function — does not know the server's release version.
 *
 * Rules:
 *  - `caps` is `null`/`undefined`/not an object → block. This client requires
 *    the caps-advertised `/dmux` transport and cannot safely talk to older
 *    servers that do not advertise caps.
 *  - `caps` is an object but lacks a mandatory cap → block
 *    `"server-incompatible"`.
 *  - mandatory or server-required cap value not in `REQUIRED_CAPS[name]` →
 *    block `"server-incompatible"`.
 *  - cap name listed in the server's `requiredCaps` but unknown to this client
 *    → block `"client-too-old"`.
 *  - unknown cap name NOT in `requiredCaps` → ignored (forward-compatible
 *    additive surfaces).
 */
export function checkServerCaps(caps: unknown, requiredCaps?: unknown): CompatibilityResult {
  // This client always uses /dmux, so a server that does not advertise caps is
  // too old to prove it supports the required sync transport.
  if (caps === null || caps === undefined || typeof caps !== "object" || Array.isArray(caps)) {
    return {
      ok: false,
      reason: "server-incompatible",
      detail: "server did not advertise compatibility caps",
    };
  }
  const capsMap = caps as Record<string, unknown>;

  for (const name of MANDATORY_CAPS) {
    if (!(name in capsMap)) {
      return {
        ok: false,
        reason: "server-incompatible",
        detail: `server is missing required cap "${name}"`,
      };
    }
    const advertised = capsMap[name];
    if (typeof advertised !== "string") {
      return {
        ok: false,
        reason: "server-incompatible",
        detail: `server advertised non-string value for cap "${name}"`,
      };
    }
    if (!REQUIRED_CAPS[name].includes(advertised)) {
      return {
        ok: false,
        reason: "server-incompatible",
        detail: `server advertised ${name}=${JSON.stringify(advertised)}, client accepts ${JSON.stringify(
          REQUIRED_CAPS[name],
        )}`,
      };
    }
  }

  // Server-required caps this client doesn't know → client too old.
  if (Array.isArray(requiredCaps)) {
    for (const name of requiredCaps) {
      if (typeof name !== "string") continue;
      if (!(CAP_NAMES as readonly string[]).includes(name)) {
        return {
          ok: false,
          reason: "client-too-old",
          detail: `server requires cap "${name}" which this client does not know`,
        };
      }
      const capName = name as CapName;
      if (!serverSupportsCapability(capsMap, capName)) {
        return {
          ok: false,
          reason: "server-incompatible",
          detail: `server requires unsupported cap "${name}"`,
        };
      }
    }
  }

  return { ok: true };
}

/**
 * Thrown by `Auth.serverInfoChecked` when `checkServerCaps` fails. Distinct
 * from `AuthError` — compatibility is not session expiry. Callers (notably
 * `maybeStartSync`) catch this specifically to hard-block sync while still
 * tolerating network/offline errors.
 */
export class CompatibilityError extends Error {
  readonly reason: "server-incompatible" | "client-too-old";
  readonly serverVersion?: string;

  constructor(
    reason: "server-incompatible" | "client-too-old",
    detail: string,
    serverVersion?: string,
  ) {
    super(detail);
    this.name = "CompatibilityError";
    this.reason = reason;
    this.serverVersion = serverVersion;
  }
}

File: server/src/caps.rs
//! Named capability versioning.
//!
//! Each constant here pins the wire/protocol shape of one compatibility surface
//! between the server and its clients. A cap value is an opaque string bumped
//! **only** on a wire-incompatible change to that surface — adding an optional
//! request/response field does NOT bump; removing or renaming a field, changing
//! a type, or changing semantics DOES bump.
//!
//! These values are advertised on `GET /api/server-info` as `caps`, plus a
//! `requiredCaps` list of cap names that clients must understand. See
//! `AGENTS.md` "Compatibility & versioning" for the full rules.

/// REST API surface (`/api/*` request/response shapes), plus the sync transport
/// the plugin now requires: Yjs documents are multiplexed over bounded `/dmux`
/// WebSocket shards, so a client on this cap will not connect to a server that
/// lacks that route.
pub const REST_API: &str = "3";

/// OAuth 2.1 server metadata + token endpoint shape, as consumed by the
/// Obsidian plugin's own OAuth flow. External MCP clients consume the OAuth
/// metadata document directly and do not see this cap.
pub const OAUTH: &str = "1";

/// cr-sqlite plugin-database replication wire format. Mirrors the client-side
/// `SYNC_FORMAT` constant in `src/pluginDb/types.ts`; the two must stay in
/// lockstep.
pub const PLUGIN_DB_SYNC: &str = "crsqlite-1";

/// Attachment shim text format committed into git backups for oversized blobs.
/// The full URI is the cap value — there is one source of truth for this
/// string, re-exported by `git::ATTACHMENT_SHIM_VERSION`.
pub const ATTACHMENT_SHIM: &str = "https://realtime.md/attachment-shim/v1";

/// Wire-visible logical replacement of long-lived Yjs documents. Clients must
/// persist and acknowledge the proposed epoch before reconnecting with a fresh
/// Y.Doc; old-epoch writes are rejected.
pub const DOCUMENT_EPOCH: &str = "1";

/// Advisory child-document invalidation messages delivered over a live vault
/// index connection. Mobile clients use this before evicting per-file Y.Docs.
pub const DOCUMENT_INVALIDATION: &str = "1";

/// Cap names clients must understand before connecting. Invalidation remains
/// optional because older clients never evict documents and can ignore it.
pub const REQUIRED: &[&str] = &["documentEpoch"];

/// All caps advertised on `/api/server-info`, in stable order.
pub fn caps() -> Vec<(&'static str, &'static str)> {
    vec![
        ("restApi", REST_API),
        ("oauth", OAUTH),
        ("pluginDbSync", PLUGIN_DB_SYNC),
        ("attachmentShim", ATTACHMENT_SHIM),
        ("documentEpoch", DOCUMENT_EPOCH),
        ("documentInvalidation", DOCUMENT_INVALIDATION),
    ]
}
